# Exploring Meteorite Data Exercise

### Prompt: 
Attached are data on all recorded meteorite landings on Earth - ever. Explore these data in any way you wish and submit a brief report outlining: 
1. The story these data appear, at first, to tell.
2. Issues that these data present. 
There is no particular right answer for this exercise. This is meant to be an exploration. Have fun with it!

## Set up the Environment

In [1]:
import pandas as pd
import numpy as np

## Import the data

In [2]:
data = pd.read_csv("C:\\Users\\d_bar\\Desktop\\meteoritessize.csv")

## Take a quick look at the data

In [3]:
data.head(10)

place type_of_meteorite  mass_g fell_found    year  \
0               Ur              Iron    0.00      Found -2500.0   
1         A-881561                H4  836.87      Found  1988.0   
2  Wietrzno-Bobrka              Iron  376.00      Found  -600.0   
3         A-881448                H4  341.27      Found  1988.0   
4         A-881813                L6  296.96      Found  1988.0   
5         A-881637                L3  195.62      Found  1988.0   
6         A-881993                L6  144.89      Found  1988.0   
7         A-881363                L4  105.42      Found  1988.0   
8           A-9034                L6   75.15      Found  1990.0   
9         A-881246                H4   57.07      Found  1988.0   

                                            database  coordinate_1  \
0  http://www.lpi.usra.edu/meteor/metbull.php?cod...      30.90000   
1  http://www.lpi.usra.edu/meteor/metbull.php?cod...     -72.00000   
2  http://www.lpi.usra.edu/meteor/metbull.php?cod...      49.41667   
3  http://www.lpi.usra.edu/meteor/metbull.php?cod...     -72.00000   
4  http://www.lpi.usra.edu/meteor/metbull.php?cod...     -72.00000   
5  http://www.lpi.usra.edu/meteor/metbull.php?cod...     -72.00000   
6  http://www.lpi.usra.edu/meteor/metbull.php?cod...     -72.00000   
7  http://www.lpi.usra.edu/meteor/metbull.php?cod...     -72.00000   
8  http://www.lpi.usra.edu/meteor/metbull.php?cod...     -72.00000   
9  http://www.lpi.usra.edu/meteor/metbull.php?cod...     -72.00000   

   coordinates_2  cartodb_id                  created_at  \
0       46.01667       34242  2013-02-15 20:56:02.826333   
1       26.00000        4141  2013-02-15 20:56:02.826333   
2       21.70000        6037  2013-02-15 20:56:02.826333   
3       26.00000        6340  2013-02-15 20:56:02.826333   
4       26.00000        6753  2013-02-15 20:56:02.826333   
5       26.00000        8147  2013-02-15 20:56:02.826333   
6       26.00000        9211  2013-02-15 20:56:02.826333   
7       26.00000       10471  2013-02-15 20:56:02.826333   
8       26.00000       11885  2013-02-15 20:56:02.826333   
9       26.00000       13071  2013-02-15 20:56:02.826333   

                   updated_at  \
0  2013-02-15 21:04:45.399478   
1  2013-02-15 21:11:51.900085   
2  2013-02-15 21:04:45.399478   
3  2013-02-15 21:11:51.900085   
4  2013-02-15 21:11:51.900085   
5  2013-02-15 21:11:51.900085   
6  2013-02-15 21:11:51.900085   
7  2013-02-15 21:11:51.900085   
8  2013-02-15 21:11:51.900085   
9  2013-02-15 21:11:51.900085   

                                            the_geom            year_date  
0  0101000020E61000008369183E220247406666666666E6...                  NaN  
1  0101000020E61000000000000000003A40000000000000...  1988-01-01 00:00:00  
2  0101000020E61000003333333333B33540B79C4B7155B5...                  NaN  
3  0101000020E61000000000000000003A40000000000000...  1988-01-01 00:00:00  
4  0101000020E61000000000000000003A40000000000000...  1988-01-01 00:00:00  
5  0101000020E61000000000000000003A40000000000000...  1988-01-01 00:00:00  
6  0101000020E61000000000000000003A40000000000000...  1988-01-01 00:00:00  
7  0101000020E61000000000000000003A40000000000000...  1988-01-01 00:00:00  
8  0101000020E61000000000000000003A40000000000000...  1990-01-01 00:00:00  
9  0101000020E61000000000000000003A40000000000000...  1988-01-01 00:00:00

In [4]:
data.tail(10)

place type_of_meteorite  mass_g fell_found    year  \
34503  ��st 056         Relict OC     0.0      Found  2008.0   
34504  ��st 057         Relict OC     0.0      Found  2009.0   
34505  ��st 058         Relict OC     0.0      Found  2009.0   
34506  ��st 059         Relict OC     0.0      Found  2009.0   
34507  ��st 060         Relict OC     0.0      Found  2009.0   
34508  ��st 061         Relict OC     0.0      Found  2009.0   
34509  ��st 062         Relict OC     0.0      Found  2010.0   
34510  ��st 063         Relict OC     0.0      Found  2010.0   
34511  ��st 064         Relict OC     0.0      Found  2011.0   
34512  ��st 049         Relict OC     0.0      Found  2012.0   

                                                database  coordinate_1  \
34503  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   
34504  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   
34505  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   
34506  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   
34507  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   
34508  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   
34509  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   
34510  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   
34511  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   
34512  http://www.lpi.usra.edu/meteor/metbull.php?cod...      58.58333   

       coordinates_2  cartodb_id                  created_at  \
34503       13.43333       34475  2013-02-15 20:56:02.826333   
34504       13.43333       34476  2013-02-15 20:56:02.826333   
34505       13.43333       34477  2013-02-15 20:56:02.826333   
34506       13.43333       34478  2013-02-15 20:56:02.826333   
34507       13.43333       34479  2013-02-15 20:56:02.826333   
34508       13.43333       34480  2013-02-15 20:56:02.826333   
34509       13.43333       34481  2013-02-15 20:56:02.826333   
34510       13.43333       34482  2013-02-15 20:56:02.826333   
34511       13.43333       34483  2013-02-15 20:56:02.826333   
34512       13.43333       34484  2013-02-15 20:56:02.826333   

                       updated_at  \
34503  2013-02-15 21:11:51.900085   
34504  2013-02-15 21:11:51.900085   
34505  2013-02-15 21:11:51.900085   
34506  2013-02-15 21:11:51.900085   
34507  2013-02-15 21:11:51.900085   
34508  2013-02-15 21:11:51.900085   
34509  2013-02-15 21:11:51.900085   
34510  2013-02-15 21:11:51.900085   
34511  2013-02-15 21:11:51.900085   
34512  2013-02-15 21:11:51.900085   

                                                the_geom            year_date  
34503  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2008-01-01 00:00:00  
34504  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2009-01-01 00:00:00  
34505  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2009-01-01 00:00:00  
34506  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2009-01-01 00:00:00  
34507  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2009-01-01 00:00:00  
34508  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2009-01-01 00:00:00  
34509  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2010-01-01 00:00:00  
34510  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2010-01-01 00:00:00  
34511  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2011-01-01 00:00:00  
34512  0101000020E610000059C0046EDDDD2A404963B48EAA4A...  2012-01-01 00:00:00

- We have roughly 34,513 records with 13 attributes each.
- `place` is likely a human-readable name for where the meteorite was found.  When `place` has a name like `A-881561` it probably means that it's from Antarctica.  
- There are some characters that were replaced with the unicode replacement character �, so cleaning here would be needed.
- `mass_g` is the mass of the meteorite in grams.
- We notice some redundancy with the `year` and `year_date` attributes.
- `cartodb_id` appears to be an id for these records in a database that came from software called Carto which is geographic information systems software.
- `created` and `updated` are the dates when the records were created in the database and when they were last updated.
- The `fell_found` attribute probably tells us if the meteorite was actually observed falling or just found.
- I'm not exactly sure what `the_geom` is, but my guess is that it's some kind of code used in GIS software which has information about the geometry of that particular part of the Earth.  
- We also notice that `coordinate_1` and `coordinates_2` have slightly different names which could become annoying.

## What questions can we ask about the data?

What are the different types of meteorites?

In [5]:
data['type_of_meteorite'].value_counts()

L6                 6565
H5                 5620
H4                 3322
H6                 3230
L5                 2732
                   ... 
LL4/5                 1
L3.7-3.9              1
Eucrite-an            1
C4/5                  1
L6-melt breccia       1
Name: type_of_meteorite, Length: 398, dtype: int64

There's 398 different types of meteorites which seems like a lot.  How many of them only appear once?

In [6]:
counts = data['type_of_meteorite'].value_counts() == 1

In [7]:
count = 0
for i in range(len(counts)):
    if counts[i] == True:
        count+=1
        

In [8]:
count

134

134 types of meteorites only appear once in the dataset.  This many unique values could pose an issue.  Perhaps there's a way to cluster the types to reduce the number of unique values.

How many meteorites were found vs observed falling?

In [9]:
data['fell_found'].value_counts()

Found                   33148
Fell                     1189
Age <70 Ma                  4
Age <300 Ma                 4
Age <500 Ma                 3
                        ...  
Age 73.3 +/- 5.3 Ma         1
Age 500 +/- 20 Ma           1
Age 0.25 +/- 0.05 Ma        1
Age >160 +/- 10 Ma          1
Age 1.07 Ma                 1
Name: fell_found, Length: 155, dtype: int64

We see that 33,148 meteorites were found while only 1,189 were observed falling.  But those add up to only 34337 and we know that there are 34513 records.  The missing 176 records seem to be taken up by other data approximating how long ago the meteor fell.  This could be cleaned by adding another column called 'age' then moving the 176 records to that column and replacing the original values with 'found'.

Are there any missing values in the data?

In [10]:
data.isna().sum()

place                  0
type_of_meteorite      0
mass_g                 0
fell_found             0
year                 347
database               0
coordinate_1           0
coordinates_2          0
cartodb_id             0
created_at             0
updated_at             0
the_geom               1
year_date            351
dtype: int64

We see 347 missing records for `year`, 351 for `year_date`, and 1 for `the_geom`. 

We notice that the likely reason for `year_date` attribute having records with `NaN` is because `year` can be a negative number.  This might require some cleaning.

What is the record with no entry for `the_geom`?

In [11]:
geom = data['the_geom'].isna()

In [12]:
for i in range(len(geom)):
    if geom[i] == True:
        print(i)

24489


In [13]:
data.iloc[24489]

place                                                 Meridiani Planum
type_of_meteorite                                    Iron, IAB complex
mass_g                                                             0.0
fell_found                                                       Found
year                                                            2005.0
database             http://www.lpi.usra.edu/meteor/metbull.php?cod...
coordinate_1                                                  -1.94617
coordinates_2                                                354.47333
cartodb_id                                                       34470
created_at                                  2013-02-15 20:56:02.826333
updated_at                                  2013-02-15 21:11:51.900085
the_geom                                                           NaN
year_date                                          2005-01-01 00:00:00
Name: 24489, dtype: object

These are strange coordinates.  After looking at the record in the actual database, this record appears to be from Mars!  

Here's a quote: \
"While exploring the remnants of the heat shield on Meridiani Planum, the Mars Exploration Rover (MER) Opportunity imaged the surroundings with its panoramic camera (PanCam). On sol 324, a close-by object with a maximum dimension of 31 cm across became for the first time clearly visible in one of these images. Subsequent PanCam images reveal a smooth rock surface covered by depressions partly reminiscent of regmaglypts."

Let's look at the most common coordinates.

In [14]:
data[['coordinate_1','coordinates_2']].median()

coordinate_1    -72.00000
coordinates_2    54.88417
dtype: float64

A quick Google search shows that these coordinates are in Anarctica.  I learned in an astronomy class I took that one reason for this is because snow and ice make it easier to find meteorites because they leave impressions in the snow.  This can affect our data because it doesn't necessarily mean that there are more meteorites in Antarctica, just that they're easier to find there.

What about the masses of the meteorites?

In [15]:
data['mass_g'].describe()

count    3.451300e+04
mean     1.686183e+04
std      6.600311e+05
min      0.000000e+00
25%      5.100000e+00
50%      2.419000e+01
75%      1.706900e+02
max      6.000000e+07
Name: mass_g, dtype: float64

The largest mass is 60,000kg, and the smallest is 0kg.  The mean is 16.86183kg.  Let's take a closer look at the these records.

In [16]:
data.loc[data['mass_g'].idxmin()]

place                                                               Ur
type_of_meteorite                                                 Iron
mass_g                                                             0.0
fell_found                                                       Found
year                                                           -2500.0
database             http://www.lpi.usra.edu/meteor/metbull.php?cod...
coordinate_1                                                      30.9
coordinates_2                                                 46.01667
cartodb_id                                                       34242
created_at                                  2013-02-15 20:56:02.826333
updated_at                                  2013-02-15 21:04:45.399478
the_geom             0101000020E61000008369183E220247406666666666E6...
year_date                                                          NaN
Name: 0, dtype: object

A meteorite with no mass is also one of the oldest in the database, and was found in modern day Iraq over 4000 years ago!  

Are there other records with no recorded mass?

In [17]:
no_mass = data['mass_g'] == 0

In [18]:
zero_mass = 0
for i in range(len(no_mass)):
    if no_mass[i] == True:
        zero_mass += 1
print(zero_mass)

448


There are 448 records with no measured mass.

What about the meteorite with the largest mass?

In [19]:
data.loc[data['mass_g'].idxmax()]

place                                                             Hoba
type_of_meteorite                                            Iron, IVB
mass_g                                                      60000000.0
fell_found                                                       Found
year                                                            1920.0
database             http://www.lpi.usra.edu/meteor/metbull.php?cod...
coordinate_1                                                 -19.58333
coordinates_2                                                 17.91667
cartodb_id                                                           1
created_at                                  2013-02-15 20:56:02.826333
updated_at                                  2013-02-15 21:11:51.900085
the_geom             0101000020E61000006D3997E2AAEA314093C6681D5595...
year_date                                          1920-01-01 00:00:00
Name: 426, dtype: object

The meteorite with the largest mass was found in Namibia and is called Hoba.  The Wikipedia article on it notes that it is the most massive naturally occuring piece of iron known on the Earth's surface!

What's the meteorite with the most recent `year` ?

In [20]:
data.loc[data['year'].idxmax()]

place                                                  Battle Mountain
type_of_meteorite                                                   L6
mass_g                                                          2900.0
fell_found                                                        Fell
year                                                            2012.0
database             http://www.lpi.usra.edu/meteor/metbull.php?cod...
coordinate_1                                                  40.66813
coordinates_2                                               -117.18913
cartodb_id                                                        2300
created_at                                  2013-02-15 20:56:02.826333
updated_at                                  2013-02-15 21:11:51.900085
the_geom             0101000020E6100000522CB7B41A4C5DC0FDBCA9488555...
year_date                                          2012-01-01 00:00:00
Name: 2711, dtype: object

According to the database, this meteorite was observed falling in Nevada using the National Weather Service's NEXRAD radar network.  It goes on to discuss the relatively short time it spent in the air, "This time span of 97 s is short compared to other meteorite falls observed by radar. This could be a result of meteorite production by a single, large breakup event, by relatively little fragmentation, or a combination of the two factors."

Let's look at the summary statistics for the categorical data.

In [21]:
data.describe(include='object').T

count unique  \
place              34513  34513   
type_of_meteorite  34513    398   
fell_found         34513    155   
database           34513  34513   
created_at         34513      1   
updated_at         34513      3   
the_geom           34512  17527   
year_date          34162    312   

                                                                 top   freq  
place                                                             Ur      1  
type_of_meteorite                                                 L6   6565  
fell_found                                                     Found  33148  
database           http://www.lpi.usra.edu/meteor/metbull.php?cod...      1  
created_at                                2013-02-15 20:56:02.826333  34513  
updated_at                                2013-02-15 21:11:51.900085  34162  
the_geom           0101000020E6100000B79C4B7155D541400000000000E0...   5636  
year_date                                        1979-01-01 00:00:00   3804

We see that it's more common for meteorites to be found than observed falling.

We see that the common type of meteorite is an L6 meteorite.  According to the database, these are low-iron chondrites (stony meteorites) which have homogenous mineral compositions and didn't undergo any melting.  

We also see that this database was created in 2013 since all the records have the same `created_at` date.  The records have only been updated three times since the creation of the database since there are only 3 unique dates for `updated_at`.

The year with the most meteorites was 1979 with 3804 records from that year.

There are 17,527 unique records for `the_geom` attribute, and one appears 5,636 times.  The interpretation of this however is unclear.

There's 312 unique values for `year_date` out of 34,513 records which isn't what we expect to see.